File: src/agent/tools/registry.ts
import type { ToolHandler, ToolDefinition } from "../../types";
import { createLogger } from "../../logger";

const log = createLogger("tools");

const tools = new Map<string, ToolHandler>();

export function registerTool(handler: ToolHandler): void {
  if (tools.has(handler.definition.name)) {
    log.warn("Tool already registered, overwriting", { name: handler.definition.name });
  }
  tools.set(handler.definition.name, handler);
  log.info("Tool registered", { name: handler.definition.name });
}

export function getTool(name: string): ToolHandler | undefined {
  return tools.get(name);
}

export function getAllToolDefinitions(): ToolDefinition[] {
  return Array.from(tools.values()).map((t) => t.definition);
}

export async function executeTool(
  name: string,
  args: Record<string, unknown>,
): Promise<string> {
  const handler = tools.get(name);
  if (!handler) {
    return JSON.stringify({ error: `Tool '${name}' not found` });
  }

  try {
    const raw = await handler.execute(args);
    // Coerce to string — dynamic skills may return non-strings at runtime despite the type.
    if (typeof raw === "string") return raw;
    return JSON.stringify(raw ?? "");
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    log.error("Tool execution failed", { name, error: message });
    return JSON.stringify({ error: message });
  }
}

File: src/tools/registry.py
"""Tool registry — global map of tool handlers, matching TS tools/registry.ts."""

from __future__ import annotations

import json

from src.types import ToolDefinition, ToolHandler
from src.utils.logger import create_logger

log = create_logger("tools")

_tools: dict[str, ToolHandler] = {}


def register_tool(handler: ToolHandler) -> None:
    name = handler.definition.name
    if name in _tools:
        log.warn("Tool already registered, overwriting", {"name": name})
    _tools[name] = handler
    log.info("Tool registered", {"name": name})


def get_tool(name: str) -> ToolHandler | None:
    return _tools.get(name)


def get_all_tool_definitions() -> list[ToolDefinition]:
    return [h.definition for h in _tools.values()]


async def execute_tool(name: str, args: dict) -> str:
    handler = _tools.get(name)
    if not handler:
        return json.dumps({"error": f"Tool '{name}' not found"})

    try:
        raw = await handler.execute(args)
        if isinstance(raw, str):
            return raw
        return json.dumps(raw or "")
    except Exception as e:
        log.error("Tool execution failed", {"name": name, "error": str(e)})
        return json.dumps({"error": str(e)})
